### PR TITLE
Add 2 new keyboard types and infer keyboardType from autofill hints

### LINKED ISSUE #41508: Have TextInputType.datetime use UIKeyboardType.numbersAndPunctuation on iOS
We are implementing the new [text input date picker for Material design](https://material.io/components/pickers/#mobile-pickers) which just uses a text input field for the user to enter the date. We set the `keyboardType` parameter on the `TextField` to `TextInputType.datetime` to bring up a keyboard appropriate for entering a date. On Android this works well. On iOS however, it just brings up the default keyboard. This works, but it requires the user to switch to numeric mode on the default keyboard to enter the date.

From the iOS docs, it looks like they support a default keyboard that comes up in numeric mode: `UIKeyboardType.numbersAndPunctuation`. Can we switch the Flutter implementation of `TextInputType.datetime` to use this on iOS? This would seem to be a closer match and give the user a better experience.

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
## Description

- Add `TextInputType.address` and `TextInputType.name`.
- Infer `keyboardType` from `AutofillHints` if left unspecified. 
- Fix docs for autofillHints

engine PR: https://github.com/flutter/engine/pull/18202

## Related Issues
Fixes https://github.com/flutter/flutter/issues/41508

## Tests

I added the following tests:
- Infer keyboardType from autofillHints

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 1]
```diff
@@ -1105,10 +1105,162 @@ class EditableText extends StatefulWidget {
...
+  ///
+  /// To provide the best user experience and ensure your app fully supports
+  /// password autofill on iOS, follow these steps:
+  ///
+  /// - Set up your iOS app's
+  ///   [associated domains](https://developer.apple.com/documentation/safariservices/supporting_associated_domains_in_your_app).
+  /// - Some autofill hints only work with specific [keyboardType]s. For example,
+  ///   [AutofillHints.name] requires [TextInputType.name] and [AutofillHints.email]
+  ///   works only with [TextInputType.email]. Make sure the input field has a
+  ///   compatible [keyboardType].
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:1118]: I think the 2nd bullet point needs better documentation on which keyboard types and autofill hints are compatible. I did some tests on an iOS 13.5 device using a native app, most `UITextContentType` only works with `UIKeyboardType.namePhonePad` (which is mapped to `TextInputType.name` in this PR), including content types like `.streetAddress`, `.zipCode` and `.telephoneNumber` (despite there're more fitting keyboard types). I can't really make sense of it so I'm not sure we should put that behavior in public documentation.

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 2]
```diff
@@ -1105,10 +1105,162 @@ class EditableText extends StatefulWidget {
...
+    // The entries with "autofill not working" comments are the iOS text content
+    // types that should work with the specified keyboard type but won't trigger
+    // (even within a native app). Tested on iOS 13.5.
+    if (!kIsWeb) {
+      switch (defaultTargetPlatform) {
+        case TargetPlatform.iOS:
+        case TargetPlatform.macOS:
+          const Map<String, TextInputType> iOSKeyboardType = <String, TextInputType> {
+            AutofillHints.addressCity : TextInputType.name,
+            AutofillHints.addressCityAndState : TextInputType.name, // Autofill not working.
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/widgets/editable_text.dart:1150]: There're surprisingly many `UITextContentType`s that don't trigger autofill (at least I couldn't find a way to, I had added all the fields, middle name, name prefix, etc., in Contacts). Maybe I should rename the field to `textContentType`?

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 3]
```diff
@@ -263,6 +263,138 @@ void main() {
...
+          ),
+        );
+
+        await tester.tap(find.byType(EditableText));
+        await tester.showKeyboard(find.byType(EditableText));
+        controller.text = 'test';
+        await tester.idle();
+        expect(tester.testTextInput.editingState['text'], equals('test'));
+        expect(tester.testTextInput.setClientArgs['inputType']['name'], equals('TextInputType.address'));
+      });
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:329]: Will these tests work even when they're run on other platforms after the PR is merged?  I would have expected you to need to use TargetPlatformVariant, but maybe I'm wrong.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (justinmc) [reply packages/flutter/lib/src/widgets/editable_text.dart:1118]: Maybe our users should be seeking out native iOS documentation on this stuff anyway, and we should just be clear about how we map to native.

**Review by justinmc — APPROVED** [review]
LGTM

We should keep an eye on our issues after this is merged to see how users are understanding this.

File: packages/flutter/test/widgets/editable_text_test.dart (hunk 3 above)
Comment (LongCatIsLooong) [reply packages/flutter/test/widgets/editable_text_test.dart:329]: From the looks of it,  they do the same thing. Changed to `TargetPlatformVariant` as it seems to be more conventional.

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (LongCatIsLooong) [reply packages/flutter/lib/src/widgets/editable_text.dart:1118]: Unfortunately the behavior isn't documented there either. Maybe I should change the name `autofillHints` to `contentType` to emphasize that it's not only for autofill, and it doesn't guarantee the platform autofill will trigger?

File: packages/flutter/lib/src/services/autofill.dart [hunk 4]
```diff
@@ -649,9 +649,9 @@ class AutofillConfiguration {
   /// {@template flutter.services.autofill.autofillHints}
   /// For the best results, hint strings need to be understood by the platform's
   /// autofill service. The common values of hint strings can be found in
-  /// [AutofillHints], as well as the platforms that understand each of them.
+  /// [AutofillHints], as well as their availability on different platforms.
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/services/autofill.dart:652]: looking for a better word for "availability"

Comment (HansMuller) [reply packages/flutter/lib/src/services/autofill.dart:652]: I think availability is pretty good choice here. Alternatively you could say "how well they're supported on different platforms".

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (dla-kirito) [reply packages/flutter/lib/src/widgets/editable_text.dart:1151]: mm】

Comment (dla-kirito) [reply packages/flutter/lib/src/widgets/editable_text.dart:1151]: 👩

**Comment by dla-kirito** [comment]
。。。

**Comment by dla-kirito** [comment]
1

**Comment by dla-kirito** [comment]
> 1

1

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 5]
```diff
@@ -591,6 +591,7 @@ class CupertinoTextField extends StatefulWidget {
   final GestureTapCallback onTap;
 
   /// {@macro flutter.widgets.editableText.autofillHints}
+  /// {@macro flutter.services.autofill.autofillHints}
```
Comment (dla-kirito) [inline packages/flutter/lib/src/cupertino/text_field.dart:594]: Nothing to preview.

**Review by dla-kirito — COMMENTED** [review]
Nothing to preview.

File: packages/flutter/lib/src/services/autofill.dart [hunk 6]
```diff
@@ -7,9 +7,9 @@ import 'text_input.dart';
 
 /// A collection of commonly used autofill hint strings on different platforms.
 ///
```
Comment (dla-kirito) [inline packages/flutter/lib/src/services/autofill.dart:9]: Nothing to preview.1

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 7]
```diff
@@ -263,6 +263,134 @@ void main() {
...
+
+        await tester.tap(find.byType(EditableText));
+        await tester.showKeyboard(find.byType(EditableText));
+        controller.text = 'test';
+        await tester.idle();
+        expect(tester.testTextInput.editingState['text'], equals('test'));
+        expect(tester.testTextInput.setClientArgs['inputType']['name'], equals('TextInputType.text'));
+    });
+  });
+
```
Comment (dla-kirito) [inline packages/flutter/test/widgets/editable_text_test.dart:393]: 1

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 1 above)
Comment (dla-kirito) [inline packages/flutter/lib/src/widgets/editable_text.dart]: 1

File: packages/flutter/lib/src/services/autofill.dart (hunk 4 above)
Comment (dla-kirito) [reply packages/flutter/lib/src/services/autofill.dart:652]: 111

Comment (dla-kirito) [reply packages/flutter/lib/src/services/autofill.dart:652]: 111

Comment (dla-kirito) [reply packages/flutter/lib/src/services/autofill.dart:652]: 11111

File: packages/flutter/lib/src/widgets/editable_text.dart [hunk 8]
```diff
@@ -1105,10 +1106,162 @@ class EditableText extends StatefulWidget {
...
```
Comment (dla-kirito) [inline packages/flutter/lib/src/widgets/editable_text.dart:1151]: test

File: packages/flutter/test/widgets/editable_text_test.dart (hunk 3 above)
Comment (dla-kirito) [reply packages/flutter/test/widgets/editable_text_test.dart:327]: > ` Some autofill hints only work with specific [keyboardType]s. For example,
  ///   [AutofillHints.name] requires [TextInputType.name] and [AutofillHints.email]
  ///   works only with [TextInputType.email]. Make sure the input field has a
  ///   compatible [keyboardType].`

Comment (dla-kirito) [reply packages/flutter/test/widgets/editable_text_test.dart:327]:
```suggestion
      });1
```

File: packages/flutter/lib/src/material/text_field.dart [hunk 9]
```diff
@@ -722,6 +722,7 @@ class TextField extends StatefulWidget {
   final ScrollController scrollController;
 
   /// {@macro flutter.widgets.editableText.autofillHints}
+  /// {@macro flutter.services.autofill.autofillHints}
   final Iterable<String> autofillHints;
```
Comment (dla-kirito) [inline packages/flutter/lib/src/material/text_field.dart:726]: test

File: packages/flutter/lib/src/services/autofill.dart [hunk 10]
```diff
@@ -7,9 +7,9 @@ import 'text_input.dart';
...
-/// Each hint may not be supported on every platform, and may get translated to
-/// different strings on different platforms. Please refer to their documentation
```
Comment (dla-kirito) [inline packages/flutter/lib/src/services/autofill.dart:11]: test

File: packages/flutter/lib/src/material/text_field.dart (hunk 9 above)
Comment (dla-kirito) [reply packages/flutter/lib/src/material/text_field.dart:726]: test

Comment (dla-kirito) [reply packages/flutter/lib/src/material/text_field.dart:726]: 1111

File: packages/flutter/lib/src/services/autofill.dart (hunk 10 above)
Comment (dla-kirito) [reply packages/flutter/lib/src/services/autofill.dart:11]: 2222

Comment (dla-kirito) [reply packages/flutter/lib/src/services/autofill.dart:11]: 4444

File: packages/flutter/lib/src/services/autofill.dart [hunk 11]
```diff
@@ -350,7 +350,7 @@ class AutofillHints {
   /// * Otherwise, the hint string will be used as-is.
```
Comment (dla-kirito) [inline packages/flutter/lib/src/services/autofill.dart:350]: 3333

File: packages/flutter/lib/src/widgets/editable_text.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/lib/src/widgets/editable_text.dart:1151]: Not sure what `textContentType` field you're referring to. Calling the enum that's used to select a keyboard `TextInputType` seems reasonable to me.

**Review by HansMuller — APPROVED** [review]
LGTM

**Pull request merged**